### ISSUE TITLE
Discrepancies with numpy.roots

### ISSUE THREAD

**Issue by ekiefl** [NKrvavica/fqs#1]
First of all, thanks for creating this repo with great documentation and explanation of your methodology.

I'm here because I want faster quartic root calculations than provided by `numpy.roots`. Specifically, I'm looking for _real_ roots. `numpy.roots` seems to be more discriminatory in rounding imaginary parts to 0 `fqs.single_quartic`. Here are some examples:

Sometimes the imaginary part in `fqs` is not _quite_ 0. `numpy` seems to be more discriminatory at rounding imaginary parts to 0:

```
coeffs: (0.002401, -0.008925317977280052, 0.009300185690512571, -0.0018690559771812293, -0.0031393414188385284)
numpy_roots:
[ 2.27719581  +0.j            0.9293334004+0.7129046843j
  0.9293334004-0.7129046843j -0.4185290093+0.j          ]
fqs_roots:
[0.120476186 -0.0000000305j 1.7381906147+0.0000000156j
 0.120476186 +0.0000000305j 1.7381906147-0.0000000156j]
```

When values are small, differences become larger:

```
coeffs: (3.8416000000000015, -0.11310361300182059, 0.2613098019282549, -0.003289237724549378, -4.336808689942018e-19)
numpy_roots:
[ 0.0084074414+0.2602654036j  0.0084074414-0.2602654036j
  0.0126269158+0.j           -0.          +0.j          ]
fqs_roots:
[0.          +0.j           0.0126269158+0.j
 0.0084074414+0.2602654036j 0.0084074414-0.2602654036j]
```

But sometimes, values are _not_ that small, and the answers vary considerably.

```
coeffs: (0.002401, -0.008925317980992227, 0.009300185700862142, -0.0018690559843707656, -0.0031393414181160877)
numpy_roots:
[ 2.2771958104+0.j            0.9293334008+0.7129046843j
  0.9293334008-0.7129046843j -0.4185290089+0.j          ]
fqs_roots:
[ 0.9293334008+0.8349352807j  0.9293334008-0.8349352807j
 -0.4868646466+0.j            2.3455314482-0.j
```

I am expecting differences between `numpy.roots` and `fqs.single_quartic`, but unfortunately `numpy.roots`  works perfectly for my application, and `fqs.single_quartic` does not. For example, here is using `numpy.roots`:

![roots](https://user-images.githubusercontent.com/8688665/145289790-3bd96141-5f90-4ec6-ab49-4fd24e378bc8.gif)

And here is using `fqs.single_quartic`:

![fqs](https://user-images.githubusercontent.com/8688665/145289812-aab509be-cc32-4b94-bb7e-cf622afd336e.gif)

The balls intersect each other because the time until their collisions (quartic polynomial) is off somehow.

Do you have any advice or insight into why `numpy.roots` would be behaving consistently (never any ball intersections) even though its not an analytical method?

**Comment by NKrvavica** [comment]
Yes, unfortunately these are examples when closed-form solution has some issues with accuracy.
The problem is that in the process of solving the roots, there is one parameter that becomes very small (if I remember correctly it is one of the coefficients of the depressed quartic). That parameter is squared in the next step when solving a cubic, and the accuracy of the final roots is reduced because of the rounding error. When that same parameter/coefficient is exactly equal to zero, then we have two double roots, and you can solve the equation by using a different analytical equation (as a special case). But in your problem that parameter just becomes extremly small.
Unfortunately, I still do not have a solution for this problem.

**Comment by NKrvavica** [comment]
There are situations when a numerical solution gives a more accurate solution than the closed-form.

**Comment by ekiefl** [comment]
Thanks @NKrvavica. This was insightful.

**Comment by ekiefl** [comment]
Have you considered multiplying this parameter by some large factor, and then rescaling after the parameter plays its role?

**Comment by wnederpel** [comment]
If accuracy for solutions with small absolute value is important with the polynomial

a x^4 + b x^3 + c x^2 + d x + e = 0,

one could also solve the equivalent problem 

e y^4 + d y^3 + y^2 + y + a = 0
y = 1/x

which should be more accurate.
